engine/class_modules/sc_demon_hunter.cpp

```diff
@@ -4723,6 +4723,49 @@ struct vengeful_retreat_t : public demon_hunter_spell_t
   }
 };
 
+// Soul Carver =========================================================
+struct soul_carver_t : public demon_hunter_attack_t
+{
+  struct soul_carver_oh_t : public demon_hunter_attack_t
+  {
+    soul_carver_oh_t( demon_hunter_t* p )
+      : demon_hunter_attack_t( "soul_carver_oh", p, p->talent.vengeance.soul_carver->effectN( 3 ).trigger() )
+    {
+      background = dual = true;
+      may_miss = may_parry = may_dodge = false;  // TOCHECK
+    }
+  };
+
+  soul_carver_oh_t* oh;
+
+  soul_carver_t( demon_hunter_t* p, util::string_view options_str )
+    : demon_hunter_attack_t( "soul_carver", p, p->talent.vengeance.soul_carver, options_str )
+  {
+    oh        = new soul_carver_oh_t( p );
+    oh->stats = stats;
+  }
+
+  void impact( action_state_t* s ) override
+  {
+    demon_hunter_attack_t::impact( s );
+
+    if ( !result_is_hit( s->result ) )
+      return;
+
+    oh->set_target( s->target );
+    oh->schedule_execute();
+
+    p()->spawn_soul_fragment( soul_fragment::LESSER, data().effectN( 3 ).base_value() );
+  }
+
+  void tick( dot_t* d ) override
+  {
+    demon_hunter_attack_t::tick( d );
+
+    p()->spawn_soul_fragment( soul_fragment::LESSER );
```

Comment:
Technically this value comes from spell data, so instead of always spawning one, I would do data().effectN( 4 ).base_value(). Since they could change it at any point.